### PR TITLE
Implemented gradle pipeline step

### PR DESCRIPTION
**Pull request by soenkekueper**
Hey,
i've now put some more work on a real gradle pipeline step. This makes it **much** easier and comfortable running a gradle build in an jenkins pieline.

You are now able to run a gradle build using a predefined gradle with just this little pipeline
`node{
    gradle gradleName: 'gradle-4.10.2', tasks: 'clean build'
}`

I've implemented this in our test environment and saved a lot of code in our pipeline script for handling environment variables like PATH and JAVA_HOME catching stderr of gradle logs and so on.

It would be great to get this in so we could use this in our productive environment.

Basically i've moved the main logic of executing gradle into GradleExecution.java so it can now be used by pipeline and freestyle builds as well.

### CODE REVIEW

File: src/main/java/hudson/plugins/gradle/GradleExecution.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+import hudson.util.ArgumentListBuilder;
+import hudson.util.VariableResolver;
+
+/**
+ * Helper for execution of an Gradle build.
+ * Encapsulates the real gradle execution independent of build type.
+ */
+public final class GradleExecution {
+
+    private String switches;
```
Comment (jvz) [inline src/main/java/hudson/plugins/gradle/GradleExecution.java]: Looks like all these fields can be final?

File: src/main/java/hudson/plugins/gradle/Gradle.java [hunk 2]
```diff
@@ -185,35 +189,6 @@ public void setPassAllAsProjectProperties(boolean passAllAsProjectProperties) {
         this.passAllAsProjectProperties = passAllAsProjectProperties;
     }
 
-    public GradleInstallation getGradle() {
```
Comment (jvz) [inline src/main/java/hudson/plugins/gradle/Gradle.java]: Removing this here won't break any API compatibility, will it?

File: src/main/java/hudson/plugins/gradle/GradleInstallation.java [hunk 3]
```diff
@@ -14,23 +24,18 @@
...
+ * 
  * @author Gregory Boissinot
  */
-public class GradleInstallation extends ToolInstallation
-        implements EnvironmentSpecific<GradleInstallation>, NodeSpecific<GradleInstallation>, Serializable {
+public class GradleInstallation 
+        extends ToolInstallation
+        implements EnvironmentSpecific<GradleInstallation>, NodeSpecific<GradleInstallation> {
 
+    private static final long serialVersionUID = -1875114350592080381L;
```
Comment (jvz) [inline src/main/java/hudson/plugins/gradle/GradleInstallation.java]: Why is this added if you removed `Serializable`? If it's still supposed to be serializable, you can keep the interface in the class declaration for explicitness.

File: src/main/java/hudson/plugins/gradle/GradleInstallationProvider.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+ * Interface for providing {@link GradleInstallation} independent of build type.
+ * 
+ * @author Sönke Küper
+ */
+public interface GradleInstallationProvider {
+
+    /**
+     * Provides all available {@link GradleInstallation}.
+     */
+    public abstract GradleInstallation[] getInstallations();
```
Comment (jvz) [inline src/main/java/hudson/plugins/gradle/GradleInstallationProvider.java]:
```suggestion
    GradleInstallation[] getInstallations();
```

`public` and `abstract` are implied by being in an interface.

File: src/main/java/hudson/plugins/gradle/GradleStep.java [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+import hudson.tools.ToolInstallation;
+import hudson.util.VariableResolver;
+import jenkins.model.Jenkins;
+
+/**
+ * The Gradle-Step for a jenkins pipeline.
+ * 
+ * Only contains the configuration, the execution is delegated to an {@link GradleExecution}.
+ * 
+ * @author Sönke Küper
```
Comment (jvz) [inline src/main/java/hudson/plugins/gradle/GradleStep.java]: This doesn't match with the copyright header you've added. Can you fix one or the other as appropriate?

File: src/main/java/hudson/plugins/gradle/GradleExecution.java (hunk 1 above)
Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/GradleExecution.java]: Yes, you are right!
I'll fix.

File: src/main/java/hudson/plugins/gradle/Gradle.java (hunk 2 above)
Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/Gradle.java]: This was only used by the config.jelly with was also adjusted to now use getGradleName insted, so this should not be an problem.

File: src/main/java/hudson/plugins/gradle/GradleInstallation.java (hunk 3 above)
Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/GradleInstallation.java]: My IDE (eclipse) told me so ;) as the superclass ToolInstallation already implements Serializable i removed the duplicate specification. Is there any guideline concerning duplicate interface specifications?

File: src/main/java/hudson/plugins/gradle/GradleStep.java (hunk 5 above)
Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/GradleStep.java]: Yes, thanks for the hint.

File: src/main/java/hudson/plugins/gradle/Gradle.java (hunk 2 above)
Comment (jvz) [reply src/main/java/hudson/plugins/gradle/Gradle.java]: Ah ok, no problem there, then!

File: src/main/java/hudson/plugins/gradle/GradleInstallation.java (hunk 3 above)
Comment (jvz) [reply src/main/java/hudson/plugins/gradle/GradleInstallation.java]: No guidelines that I can remember. It's more of a Java style thing in general.

Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/GradleInstallation.java]: Okay you are right. i've added the Serializable interface!

File: src/main/java/hudson/plugins/gradle/GradleInstallationProvider.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package hudson.plugins.gradle;
+
+/**
+ * Interface for providing {@link GradleInstallation} independent of build type.
+ * 
+ * @author Sönke Küper
```
Comment (wolfs) [inline src/main/java/hudson/plugins/gradle/GradleInstallationProvider.java]: Let's not use `@autor` tags. Git should be enough.

File: src/main/java/hudson/plugins/gradle/GradleInstallation.java [hunk 7]
```diff
@@ -68,28 +73,46 @@ public void buildEnvVars(EnvVars env) {
...
-            public String call() throws IOException {
-                File exe = getExeFile();
-                if (exe.exists()) {
-                    return exe.getPath();
-                }
-                return null;
+    public String getExecutable(final Launcher launcher) throws IOException, InterruptedException {
+        final File exe = launcher.getChannel().call(new MasterToSlaveCallable<File, IOException>() {
+
+            private static final long serialVersionUID = 5572701806697595613L;
```
Comment (wolfs) [inline src/main/java/hudson/plugins/gradle/GradleInstallation.java]: What is the `serialVersionUID` for? IIUC, Jenkins uses XStream to serialize/deserialize data, not Java serialization. is this required?

File: src/main/java/hudson/plugins/gradle/GradleInstallationProvider.java (hunk 6 above)
Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/GradleInstallationProvider.java]: Okay, i've just done it like in the other files around.. ;)

File: src/main/java/hudson/plugins/gradle/GradleInstallation.java (hunk 7 above)
Comment (soenkekueper) [reply src/main/java/hudson/plugins/gradle/GradleInstallation.java]: Yes you are right, it's not needed here. Just a habit from my daily work. I'll remove it.

**Comment by soenkekueper** [comment]
I've installed this in our test environment and it is running very well and has a lot of advantages instead of running a gradle by executing a shell or bat file. It makes it more comfortable, easier and safer to run a gradle build, because you don't have to set any environment variables in dependency to the node you are currently running on.

As i was looking for a solution to our problems i found the ticket JENKINS-44834 and existing pull request. So i first decided to finish the work started there leading to #60. Due the changes made there were prerequisite to implement a real step i've continued the work to implement this step. So in conclusion with this PR the PR #60 is obsolete.

File: build.gradle [hunk 8]
```diff
@@ -55,14 +55,21 @@ jenkinsPlugin {
 sourceCompatibility = '1.7'
 
 dependencies {
-  compile 'org.jenkins-ci.lib:dry-run-lib:0.1'
-  compileOnly 'org.jenkins-ci:symbol-annotation:1.3'
-  jenkinsPlugins 'org.jenkins-ci.plugins:structs:1.3@jar'
+  jenkinsPlugins 'org.jenkins-ci.plugins:dry-run:0.9'
+  jenkinsPlugins 'org.jenkins-ci.plugins.workflow:workflow-step-api:2.11'
```
Comment (wolfs) [inline build.gradle]: I guess `workflow-step-api` and `worflow-support` should be in `optional JenkinsPlugins?

File: build.gradle [hunk 9]
```diff
@@ -55,14 +55,21 @@ jenkinsPlugin {
...
 
   testCompile 'org.spockframework:spock-core:0.7-groovy-1.8'
-  jenkinsTest 'org.jenkins-ci.main:jenkins-test-harness:2.8@jar'
+  testCompile 'org.jenkins-ci.main:jenkins-test-harness:2.22@jar'
+  testCompile 'org.jenkins-ci.main:jenkins-test-harness-tools:2.2'
+
+  jenkinsTest 'org.jenkins-ci.plugins.workflow:workflow-cps:2.32'
+  jenkinsTest 'org.jenkins-ci.plugins.workflow:workflow-job:2.11'
+  jenkinsTest 'org.jenkins-ci.plugins.workflow:workflow-basic-steps:2.5'
+  jenkinsTest 'org.jenkins-ci.plugins.workflow:workflow-durable-task-step:2.9'
```
Comment (wolfs) [inline build.gradle]: I see that you depend on `durable-task-step`. Is this used somewhere? I would love the Gradle step to be a durable task step, i.e. when the master loses the connection to the `agent`, the build will continue and the master will re-connect to the agent and resume the operation for where it stopped.

Does this work with the current implementation? Or maybe the step should only support a nice API and the n run an `sh`/`bat` command under the covers.

File: src/main/java/hudson/plugins/gradle/GradleExecution.java [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+                break;
+            }
+        }
+        if (gradleWrapperFile == null) {
+            listener.fatalError("The Gradle wrapper has not been found in these directories: %s", Joiner.on(", ").join(possibleWrapperLocations));
+        }
+        return gradleWrapperFile;
+    }
+
+    private FilePath getNormalizedRootBuildScriptDir(Run<?, ?> build, VariableResolver<String> resolver, FilePath moduleRoot) {
```
Comment (wolfs) [inline src/main/java/hudson/plugins/gradle/GradleExecution.java]: `build` is not used anymore.

File: src/main/java/hudson/plugins/gradle/GradleExecution.java [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+        normalizedArgs = replaceWhitespaceBySpace(normalizedArgs);
+        normalizedArgs = Util.replaceMacro(normalizedArgs, resolver);
+        return normalizedArgs;
+    }
+
+    private static String replaceWhitespaceBySpace(String argument) {
+        return argument.replaceAll("[\t\r\n]+", " ");
+    }
+
+    private List<FilePath> getPossibleWrapperLocations(Run<?, ?> build, Launcher launcher, VariableResolver<String> resolver, FilePath normalizedRootBuildScriptDir, FilePath moduleRoot) throws IOException, InterruptedException {
```
Comment (wolfs) [inline src/main/java/hudson/plugins/gradle/GradleExecution.java]: `build` and `launcher` are not used anymore.

File: src/main/java/hudson/plugins/gradle/GradleInstallation.java [hunk 12]
```diff
@@ -68,28 +72,42 @@ public void buildEnvVars(EnvVars env) {
...
+    public String getExecutable(final Launcher launcher) throws IOException, InterruptedException {
+        final File exe = launcher.getChannel().call(new MasterToSlaveCallable<File, IOException>() {
+
+            @Override
+            public File call() throws IOException {
+                return getExeFile();
+            }
+        });
+
+        final boolean existis = launcher.getChannel().call(new MasterToSlaveCallable<Boolean, IOException>() {
```
Comment (wolfs) [inline src/main/java/hudson/plugins/gradle/GradleInstallation.java]: Typo: should be `exists`.

File: build.gradle [hunk 13]
```diff
@@ -55,14 +55,21 @@ jenkinsPlugin {
 sourceCompatibility = '1.7'
 
 dependencies {
-  compile 'org.jenkins-ci.lib:dry-run-lib:0.1'
-  compileOnly 'org.jenkins-ci:symbol-annotation:1.3'
-  jenkinsPlugins 'org.jenkins-ci.plugins:structs:1.3@jar'
+  jenkinsPlugins 'org.jenkins-ci.plugins:dry-run:0.9'
```
Comment (wolfs) [inline build.gradle]: The new version of the `dry-run` plugin seems to depend on the `maven-plugin`. Can we exclude the `maven-plugin` as a dependency? Something like
```
configurations {
  all {
    exclude group: 'org.jenkins-ci.main', module: 'maven-plugin'
  }
}
```

**Comment by wolfs** [comment]
Maybe an approach as shown here would be better: https://jenkins.io/blog/2016/04/21/dsl-plugins/

**Comment by wolfs** [comment]
There is also https://github.com/jenkinsci/pipeline-maven-plugin, which seems to use the `Step` API.

**Comment by wolfs** [comment]
I'll be closing the PR, given that there hasn't been any recent activity. Is the PR still interesting? If yes, please open a new one based on current `main`. Thank you!
